src/tiled/scriptdialog.cpp

```diff
@@ -0,0 +1,270 @@
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "scriptdialog.h"
+#include "scriptmanager.h"
+#include <QJSEngine>
+#include <string.h>
+#include <memory>
+
+static int leftColumnStretch = 0;
+// stretch as much as we can so that the left column looks as close to zero width as possible when there is no content
+static int rightColumnStretch = 1;
+namespace Tiled {
+
+static void deleteAllFromLayout(QLayout *layout)
+{
+    while (QLayoutItem *item = layout->takeAt(0)) {
+        delete item->widget();
+
+        if (QLayout *layout = item->layout())
+            deleteAllFromLayout(layout);
+
+        delete item;
+    }
+}
+
+ScriptDialog::ScriptDialog(const QString &title)
+ : QDialog(MainWindow::maybeInstance())
+{
+    if (title.isEmpty())
+        QDialog::setWindowTitle(tr("Script"));
+    else
+        QDialog::setWindowTitle(title);
+
+    setAttribute(Qt::WA_DeleteOnClose);
+    m_lastWidgetTypeName.clear();
+    setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+    m_gridLayout = new QGridLayout(this);
+    initializeLayout();
+}
+void ScriptDialog::initializeLayout()
+{
+    // make the right-hand column more likely to stretch
+    m_gridLayout->setColumnStretch(0, leftColumnStretch);
+    m_gridLayout->setColumnStretch(1, rightColumnStretch);
+    m_rowIndex = 0;
+
+    addNewRow();
+}
+void ScriptDialog::clear()
+{
+    deleteAllFromLayout(layout());
+    initializeLayout();
+}
+
+QLabel *ScriptDialog::addHeading(const QString &text, bool fillRow)
+{
+    addNewRow();
+    QLabel *label = newLabel(text);
+    m_gridLayout->addWidget(label, m_rowIndex, m_widgetsInRow, 1, fillRow? -1 : 1); // max width
+    label->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
+    m_widgetsInRow++;
+    if (fillRow) {
+        label->setWordWrap(true);
+        addNewRow();
+    }
+    return label;
+}
+QLabel *ScriptDialog::addLabel(const QString &text)
+{
+    moveToRightColumn();
+    return (QLabel *)addDialogWidget(newLabel(text));
+}
+
+QFrame *ScriptDialog::addSeparator(const QString &labelText)
+{
+    addNewRow();
+    m_rowLayout = new QHBoxLayout();
+    // we don't use addDialogWidget() here so that we can make the size of the separator
+    // level independent of the size of the left column.
+    m_gridLayout->addLayout(m_rowLayout, m_rowIndex, 0, 1, -1); // span entire row
+    if (!labelText.isEmpty()) {
+        QLabel *separatorLabel;
+        separatorLabel= newLabel(labelText);
+        separatorLabel->setWordWrap(false);
+        m_rowLayout->addWidget(separatorLabel, 1);
+    }
+    QFrame *line = new QFrame(this);
+    line->setFrameShape(QFrame::HLine);
+    line->setFrameShadow(QFrame::Sunken);
+    m_rowLayout->addWidget(line, rightColumnStretch); // higher stretch
+    addNewRow();
+    return line;
+}
+QLineEdit *ScriptDialog::addTextInput(const QString &labelText, const QString &defaultValue)
+{
+    checkIfSameType("QLineEdit");
+    moveToRightColumn();
```

Comment:
It seems all calls to `checkIfSameType` are followed by `moveToRightColumn`, so you might consider whether it makes sense to include `moveToRightColumn` in `checkIfSameType`.